src/nvim/buffer.c

```diff
@@ -1231,20 +1231,25 @@ do_buffer(
           return FAIL;
         }
       } else {
-        if (buf->terminal) {
-          if (p_confirm || cmdmod.confirm) {
-            if (!dialog_close_terminal(buf)) {
-              return FAIL;
-            }
-          } else {
-            EMSG2(_("E89: %s will be killed (add ! to override)"),
-                  (char *)buf->b_fname);
+        EMSGN(_("E89: No write since last change for buffer %" PRId64
+                " (add ! to override)"),
+              buf->b_fnum);
+        return FAIL;
+      }
+    }
+
+    if (!forceit && buf->terminal) {
+      uint64_t chan_id;
+      get_option_value("channel", (long *)&chan_id, NULL, OPT_LOCAL);
```

Comment:
this makes an assumption about curbuf which we shouldn't introduce here. I was hoping that `f_getbufvar` would lead to a function that takes a `buf`, but no... https://github.com/neovim/neovim/blob/8af13ed946aa06636c633327c6549993e2e50116/src/nvim/eval/funcs.c#L3004-L3024

Reply:
We can use `get_option_value_strict` instead which takes a `buf_T` as its `from` parameter (this is how `nvim_buf_get_option` does it).